File: app/src/main/java/com/xyoye/dandanplay/ui/activities/WebViewActivity.java
package com.xyoye.dandanplay.ui.activities;

import android.annotation.SuppressLint;
import android.content.Intent;
import android.graphics.Color;
import android.view.KeyEvent;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;
import android.webkit.WebChromeClient;
import android.webkit.WebSettings;
import android.webkit.WebView;
import android.webkit.WebViewClient;

import com.blankj.utilcode.util.ConvertUtils;
import com.blankj.utilcode.util.StringUtils;
import com.blankj.utilcode.util.ToastUtils;
import com.xyoye.dandanplay.R;
import com.xyoye.dandanplay.base.BaseMvcActivity;
import com.xyoye.dandanplay.ui.weight.ProgressView;

import butterknife.BindView;

/**
 * Created by xyoye on 2019/3/30.
 */

public class WebViewActivity extends BaseMvcActivity {

    @BindView(R.id.web_view)
    WebView mWebView;
    @BindView(R.id.toolbar)
    android.support.v7.widget.Toolbar toolbar;

    private ProgressView progressView;      //进度条
    private boolean isSelectUrl = false;    //选择url模式

    @Override
    protected int initPageLayoutID() {
        return R.layout.activity_webview;
    }

    @Override
    public void initPageView() {
        Intent intent = getIntent();
        String title = intent.getStringExtra("title");
        String link = intent.getStringExtra("link");
        isSelectUrl = intent.getBooleanExtra("isSelect", false);

        setTitle(title);

        initView();

        initWebSetting();

        mWebView.setWebViewClient(mWebViewClient);
        mWebView.setWebChromeClient(mWebChromeClient);
        mWebView.loadUrl(link);
    }

    @Override
    public void initPageViewListener() {

    }

    private void initView() {
        //初始化进度条
        progressView = new ProgressView(this);
        progressView.setLayoutParams(new ViewGroup.LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT, ConvertUtils.dp2px(4)));
        progressView.setColor(Color.BLUE);
        progressView.setProgress(10);
        //把进度条加到WebView中
        mWebView.addView(progressView);
    }

    @SuppressLint("SetJavaScriptEnabled")
    private void initWebSetting(){
        WebSettings webSetting = mWebView.getSettings();
        webSetting.setJavaScriptEnabled(true);
        //  WebSettings.LOAD_DEFAULT 如果本地缓存可用且没有过期则使用本地缓存，否加载网络数据 默认值
        //  WebSettings.LOAD_CACHE_ELSE_NETWORK 优先加载本地缓存数据，无论缓存是否过期
        //  WebSettings.LOAD_NO_CACHE  只加载网络数据，不加载本地缓存
        //  WebSettings.LOAD_CACHE_ONLY 只加载缓存数据，不加载网络数据
        //Tips:有网络可以使用LOAD_DEFAULT 没有网时用LOAD_CACHE_ELSE_NETWORK
        webSetting.setCacheMode(android.webkit.WebSettings.LOAD_NO_CACHE);
        webSetting.setAppCacheEnabled(true);    //开启H5(APPCache)缓存功能websettings.setAppCacheMaxSize(1024*1024*8);
        String appCachePath = this.getApplicationContext().getCacheDir().getAbsolutePath();
        webSetting.setAppCachePath(appCachePath);
        //开启 DOM storage API 功能 较大存储空间，使用简单
        webSetting.setDomStorageEnabled(true);
        //开启 Application Caches 功能 方便构建离线APP 不推荐使用
        webSetting.setAppCacheEnabled(true);
        //支持通过JS打开新窗口
        webSetting.setJavaScriptCanOpenWindowsAutomatically(true);
        //视图适应窗口
//        webSetting.setUseWideViewPort(true);
//        webSetting.setLoadWithOverviewMode(true);
        webSetting.setLayoutAlgorithm(WebSettings.LayoutAlgorithm.SINGLE_COLUMN);
        //允许 WebView 使用 File 协议
        webSetting.setAllowFileAccess(true);
        //允许webview对文件的操作
        webSetting.setAllowUniversalAccessFromFileURLs(true);
        webSetting.setAllowFileAccessFromFileURLs(true);
    }

    @Override
    public boolean onKeyDown(int keyCode, KeyEvent event) {
        if (keyCode == KeyEvent.KEYCODE_BACK) {
            if (mWebView.canGoBack()) {
                mWebView.goBack();
                return true;
            } else {
                finish();
            }
        }
        return super.onKeyDown(keyCode, event);
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        switch (item.getItemId()) {
            case android.R.id.home:
                WebViewActivity.this.finish();
                break;
            case R.id.select_url:
                if (!StringUtils.isEmpty(mWebView.getUrl())) {
                    Intent intent = getIntent();
                    intent.putExtra("selectUrl", mWebView.getUrl());
                    setResult(RESULT_OK, intent);
                    WebViewActivity.this.finish();
                } else {
                    ToastUtils.showShort("url不能为空");
                }
                break;
        }
        return super.onOptionsItemSelected(item);
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        if (isSelectUrl)
            getMenuInflater().inflate(R.menu.menu_url_select, menu);
        return super.onCreateOptionsMenu(menu);
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
        //防止webView内存泄漏
        if (mWebView != null) {
            mWebView.loadDataWithBaseURL(null, "", "text/html", "utf-8", null);
            mWebView.clearHistory();
            ((ViewGroup) mWebView.getParent()).removeView(mWebView);
            mWebView.destroy();
            mWebView = null;
        }
    }

    WebViewClient mWebViewClient = new WebViewClient() {

        @Override
        public boolean shouldOverrideUrlLoading(WebView view, String url) {
            view.loadUrl(url);
            return true;
        }
    };

    WebChromeClient mWebChromeClient = new WebChromeClient() {
        @Override
        public void onProgressChanged(WebView view, int newProgress) {
            if (newProgress == 100) {
                //加载完毕进度条消失
                progressView.setVisibility(View.GONE);
            } else {
                //更新进度
                progressView.setProgress(newProgress);
            }
            super.onProgressChanged(view, newProgress);
        }
    };
}

File: user_component/src/main/java/com/xyoye/user_component/ui/activities/web_view/WebViewActivity.kt
package com.xyoye.user_component.ui.activities.web_view

import android.annotation.SuppressLint
import android.content.Intent
import android.net.Uri
import android.view.KeyEvent
import android.view.Menu
import android.view.MenuItem
import android.view.ViewGroup
import android.webkit.WebChromeClient
import android.webkit.WebResourceRequest
import android.webkit.WebSettings
import android.webkit.WebView
import android.webkit.WebViewClient
import androidx.core.view.isGone
import com.alibaba.android.arouter.facade.annotation.Autowired
import com.alibaba.android.arouter.facade.annotation.Route
import com.alibaba.android.arouter.launcher.ARouter
import com.xyoye.common_component.base.BaseActivity
import com.xyoye.common_component.config.RouteTable
import com.xyoye.common_component.extension.toResColor
import com.xyoye.common_component.network.config.Api
import com.xyoye.common_component.utils.dp2px
import com.xyoye.user_component.BR
import com.xyoye.user_component.R
import com.xyoye.user_component.databinding.ActivityWebViewBinding
import com.xyoye.user_component.ui.weight.WebViewProgress

@SuppressLint("SetJavaScriptEnabled")
@Route(path = RouteTable.User.WebView)
class WebViewActivity : BaseActivity<WebViewViewModel, ActivityWebViewBinding>() {

    @JvmField
    @Autowired
    var titleText: String = ""

    @JvmField
    @Autowired
    var url: String? = ""

    @JvmField
    @Autowired
    var isSelectMode: Boolean = false

    private lateinit var progressView: WebViewProgress

    override fun initViewModel() =
        ViewModelInit(
            BR.viewModel,
            WebViewViewModel::class.java
        )

    override fun getLayoutId() = R.layout.activity_web_view

    override fun initView() {
        ARouter.getInstance().inject(this)

        val realUrl = url
        if (realUrl.isNullOrEmpty()) {
            finish()
            return
        }

        title = titleText

        //进度条
        progressView = WebViewProgress(this).apply {
            layoutParams = ViewGroup.LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT, dp2px(4))
            setColor(R.color.text_blue.toResColor())
            setProgress(10)
        }

        dataBinding.webView.apply {
            addView(progressView)
            settings.apply {
                allowFileAccess = true
                javaScriptEnabled = true
                domStorageEnabled = true
                cacheMode = WebSettings.LOAD_NO_CACHE
                javaScriptCanOpenWindowsAutomatically = true
                allowUniversalAccessFromFileURLs = true
            }
            webViewClient = object : WebViewClient() {

                override fun shouldOverrideUrlLoading(view: WebView?, request: WebResourceRequest?): Boolean {
                    val uri = request?.url ?: return false

                    // 选取b站视频时，考虑将APP协议地址转换为普通番剧地址
                    val bangumiUrl = considerMapBangumiUrl(uri)
                    if (bangumiUrl != null && bangumiUrl != view?.url) {
                        view?.loadUrl(bangumiUrl)
                        return true
                    }

                    // 非http(s)协议资源，不加载
                    if (uri.scheme?.startsWith("http") != true) {
                        return true
                    }

                    return false
                }
            }

            webChromeClient = object : WebChromeClient() {
                override fun onProgressChanged(view: WebView, newProgress: Int) {
                    if (newProgress == 100) {
                        //加载完毕进度条消失
                        progressView.isGone = true
                    } else {
                        //更新进度
                        progressView.setProgress(newProgress)
                    }
                    super.onProgressChanged(view, newProgress)
                }
            }
        }

        dataBinding.webView.loadUrl(realUrl)
    }

    /**
     * 将BiliBili APP协议地址转换为普通番剧地址
     */
    private fun considerMapBangumiUrl(schemeUri: Uri): String? {
        if (isSelectMode.not()) {
            return null
        }

        val isAppBangumiUrl = schemeUri.scheme == "bilibili"
                && schemeUri.host == "bangumi"
                && schemeUri.pathSegments.firstOrNull() == "season"
        if (isAppBangumiUrl.not()) {
            return null
        }

        val seasonId = schemeUri.pathSegments.lastOrNull()
        if (seasonId.isNullOrEmpty()) {
            return null
        }

        return Uri.parse(Api.BILI_BILI_M)
            .buildUpon()
            .appendPath("bangumi")
            .appendPath("play")
            .appendPath("ss$seasonId")
            .build()
            .toString()
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        if (isSelectMode) {
            menuInflater.inflate(R.menu.menu_web_view, menu)
        }
        return super.onCreateOptionsMenu(menu)
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        if (item.itemId == R.id.item_select_url) {
            val url = dataBinding.webView.url
            val intent = Intent()
            intent.putExtra("url_data", url)
            setResult(RESULT_OK, intent)
            finish()
        }
        return super.onOptionsItemSelected(item)
    }

    override fun onKeyDown(keyCode: Int, event: KeyEvent?): Boolean {
        if (keyCode == KeyEvent.KEYCODE_BACK) {
            if (dataBinding.webView.canGoBack()) {
                dataBinding.webView.goBack()
                return true
            } else {
                finish()
            }
        }
        return super.onKeyDown(keyCode, event)
    }

    override fun onDestroy() {
        super.onDestroy()
        dataBinding.webView.apply {
            loadDataWithBaseURL(null, "", "text/html", "utf-8", null)
            clearHistory()
            (parent as ViewGroup).removeView(this)
            destroy()
        }
    }
}